### PR TITLE
[#164986888] More logging for the happy path

### PR DESCRIPTION
**Pull request by 46bit**
# What

This adds some happy-path logging to the internals of the CDN Broker. These info logs should give us useful information about the broker's activity when in production. In particular, it tells us how long the broker waits on LetsEncrypt to solve the challenges.

(*Solve the challenges* means LetsEncrypt observing that the DNS records or text files are in place to prove domain ownership and so allow issuing the TLS certificate.)

## How to review

~ℹ️ *This includes a bunch of commits from https://github.com/alphagov/paas-cdn-broker/pull/15. Review that first to make things a bit easier.* ℹ️~

* See if the tests pass;
* Code review;
* Deploy in a dev env and try provisioning, deprovisioning, updating, etc.

## Who can review

Not @46bit

### CODE REVIEW

**Comment by 46bit** [comment]
I've rebased this now that https://github.com/alphagov/paas-cdn-broker/pull/15 has been merged.

File: models/models.go [hunk 1]
```diff
@@ -809,30 +809,38 @@ func (m *RouteManager) ensureChallenges(route *Route, client *acme.Client, updat
...
 		return nil
 	}
 
-	return nil
+	lsession.Info("db-save-route-challenge")
+	err = m.db.Save(route).Error
+	if err != nil {
+		lsession.Error("db-save-route-challenge-err", err)
+	}
+	return err
```
Comment (AP-Hunt) [inline models/models.go]: Could we insert a "challenge-ensure-success" message, or something of that ilk. Essentially a message we can scan for to see if everything was successful here?

File: models/models.go [hunk 2]
```diff
@@ -851,22 +851,26 @@ func (m *RouteManager) GetDNSInstructions(route *Route) ([]string, error) {
...
+	lsession.Info("load-user")
 	user, err := route.loadUser(m.db)
 	if err != nil {
-		lsession.Error("load-user", err)
+		lsession.Error("load-user-err", err)
 		return instructions, err
 	}
 
 	if err := json.Unmarshal(route.ChallengeJSON, &challenges); err != nil {
 		lsession.Error("json-unmarshal-challenge", err)
```
Comment (AP-Hunt) [inline models/models.go:890]: In [`ensureChallenges`](https://github.com/alphagov/paas-cdn-broker/pull/16/commits/643d6e56d10cefdb8d70298e813cdbeaa60689be#diff-0b230c6239e29d9787ee9d482d175728R825) we logged a message before unmarshalling. I think we should be consistent about that either way (adding here, or removing it there)

File: models/models.go [hunk 3]
```diff
@@ -851,22 +851,26 @@ func (m *RouteManager) GetDNSInstructions(route *Route) ([]string, error) {
...
+
+	lsession.Info("get-key-authorization")
 	for _, auth := range challenges {
 		for _, challenge := range auth.Body.Challenges {
 			if challenge.Type == acme.DNS01 {
+				lsession.Info("get-key-authorization-for-a-dns-challenge", lager.Data{"domain": auth.Domain})
 				keyAuth, err := acme.GetKeyAuthorization(challenge.Token, user.GetPrivateKey())
 				if err != nil {
-					lsession.Error("get-key-authorization", err)
+					lsession.Error("get-key-authorization-err", err)
```
Comment (AP-Hunt) [inline models/models.go:901]: Is it possible to include the domain in the data here too? I know these logs can sometimes be a bit dense and hard to parse, so having the information in a line single is helpful.

File: models/models.go [hunk 4]
```diff
@@ -785,22 +795,22 @@ func (m *RouteManager) deployCertificate(route Route, cert acme.CertificateResou
 		"instance-id": route.InstanceId,
 	})
 
+	lsession.Info("get-cert-expiry")
 	expires, err := acme.GetPEMCertExpiration(cert.Certificate)
 	if err != nil {
-		lsession.Error("get-cert-expiry", err)
+		lsession.Error("get-cert-expiry-err", err)
```
Comment (AP-Hunt) [inline models/models.go:802]: Can we include the domain here?

File: models/models.go [hunk 5]
```diff
@@ -785,22 +795,22 @@ func (m *RouteManager) deployCertificate(route Route, cert acme.CertificateResou
...
 		return err
 	}
 
 	name := fmt.Sprintf("cdn-route-%s-%s", route.InstanceId, expires.Format("2006-01-02_15-04-05"))
-
-	m.logger.Info("Uploading certificate to IAM", lager.Data{"name": name})
-
+	m.logger.Info("iam-upload-certificate", lager.Data{"name": name})
```
Comment (AP-Hunt) [inline models/models.go]: Same here

File: models/models.go [hunk 6]
```diff
@@ -785,22 +795,22 @@ func (m *RouteManager) deployCertificate(route Route, cert acme.CertificateResou
...
 	certId, err := m.iam.UploadCertificate(name, cert)
 	if err != nil {
-		lsession.Error("iam-upload-certificate", err)
+		lsession.Error("iam-upload-certificate-err", err)
```
Comment (AP-Hunt) [inline models/models.go]: And here?

**Review by AP-Hunt — CHANGES_REQUESTED** [review]
Looking pretty reasonable to me, and thanks for splitting it in to 3 distinct commits. Just a couple comments requesting _moar data_ in the logs

File: models/models.go (hunk 4 above)
Comment (46bit) [reply models/models.go:802]: Resolved by including the domains in all this function via the `logger.Session` call.

File: models/models.go (hunk 5 above)
Comment (46bit) [reply models/models.go]: Resolved by including the domains in all this function via the `logger.Session` call.

File: models/models.go (hunk 6 above)
Comment (46bit) [reply models/models.go]: Resolved by including the domains in all this function via the `logger.Session` call.

File: models/models.go (hunk 2 above)
Comment (46bit) [reply models/models.go:890]: Removed the message from `ensureChallenges` as it didn't have much value on top of the error-case logging.

File: models/models.go (hunk 3 above)
Comment (46bit) [reply models/models.go:901]: Resolved by including the domains in all this function via the `logger.Session` call.

File: models/models.go (hunk 1 above)
Comment (46bit) [reply models/models.go]: I added an `updated` message, and tweaked the function's other successful exit point to match.

**Comment by 46bit** [comment]
@AP-Hunt Ready for re-review :) The diff of what I've done is https://github.com/alphagov/paas-cdn-broker/compare/6c38c68c4c9761d5ee1f7497a800cbc07819d0ca..d68c2e14b2a5debbe49b3e711224dc70ce57ed7c. (It also includes #18 because I rebased against current `gds_master`.)

File: models/models.go [hunk 7]
```diff
@@ -783,55 +792,83 @@ func (m *RouteManager) solveChallenges(clients map[acme.Challenge]*acme.Client,
...
+			"err":     err,
+		})
+		return err
+	}
+	return nil
 }
 
 func (m *RouteManager) ensureChallenges(route *Route, client *acme.Client, update bool) error {
 	lsession := m.logger.Session("ensure-challenges", lager.Data{
 		"instance-id": route.InstanceId,
```
Comment (AP-Hunt) [inline models/models.go:834]: Could we include the domain here too? If we include it everywhere (ideally with the same key, if it makes sense) we can easily grep for it

**Review by AP-Hunt — CHANGES_REQUESTED** [review]
1 small change, and the tests aren't passing

File: models/models.go (hunk 7 above)
Comment (46bit) [reply models/models.go:834]: Added

**Comment by tlwr** [comment]
I reviewed #16 #17 and #19 at the same time, and did a happy path journey:

- Create
- Update
- Delete

and everything seems to work fine.
